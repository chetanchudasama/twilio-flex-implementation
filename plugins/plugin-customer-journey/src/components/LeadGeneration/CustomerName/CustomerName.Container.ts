import { connect } from "react-redux";
import { Dispatch } from "redux";
import { CustomerDetailModel } from "@common/components";

import { AppState, namespace } from "../../../states";
import { CustomerName } from "./CustomerName";
import { StateToProps, DispatchToProps } from "./CustomerName.Props";

const mapStateToProps = (state: AppState): StateToProps => {
  const customer = Object.assign(
    new CustomerDetailModel(),
    state[namespace].customer.customer
  );
  return {
    firstName: customer.firstName,
    fullName: customer.fullName,
  };
};

const mapDispatchToProps = (dispatch: Dispatch<any>): DispatchToProps => ({});

export default connect(mapStateToProps, mapDispatchToProps)(CustomerName);
