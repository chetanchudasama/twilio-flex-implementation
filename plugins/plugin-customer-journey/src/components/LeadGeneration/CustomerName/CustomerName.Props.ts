export interface StateToProps {
  firstName: string;
  fullName: string;
}

export interface DispatchToProps {}

export interface OwnProps {
  updateCustomerPreferredName: (preferredName: string) => void;
}

export type CustomerNameProps = StateToProps & DispatchToProps & OwnProps;
