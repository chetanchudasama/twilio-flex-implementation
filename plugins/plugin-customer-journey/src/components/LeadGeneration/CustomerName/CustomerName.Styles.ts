import { black, defaultSnackbarColor, theme } from "@common/components";
import { fade } from "@material-ui/core/styles/colorManipulator";
import styled from "react-emotion";

export const CustomerNameStyles = styled("div")({
  padding: "24px 24px 24px 60px",
  ".heading": {
    fontSize: 20,
    lineHeight: "24px",
    fontWeight: 600,
    color: black,
  },
  ".custom-input": {
    width: 293,
    marginTop: "0 !important",
    marginBottom: "0 !important",
  },
  ".custom-input > div": {
    marginTop: "0 !important",
    height: 40,
  },
  ".customer-name-toggle": {
    paddingBottom: "2px !important",
  },
  ".root": {
    padding: 4,
    backgroundColor: defaultSnackbarColor,
    borderRadius: 3,
    width: "fit-content",
    boxShadow: "none",
  },
  ".button": {
    "&:first-child": {
      marginLeft: 0,
    },
    "&:last-child": {
      marginRight: 0,
    },
    marginLeft: 2,
    marginRight: 2,
    fontWeight: 600,
    borderRadius: "3px !important",
    color: `${fade(theme.palette.common.black, 0.6)} !important`,
    padding: theme.spacing.unit * 4,
  },
  ".selected": {
    backgroundColor: `${theme.palette.common.white} !important`,
    color: `${theme.palette.secondary.main} !important`,
    boxShadow:
      "0px 1px 5px 0px rgb(0 0 0 / 20%), 0px 2px 2px 0px rgb(0 0 0 / 14%), 0px 3px 1px -2px rgb(0 0 0 / 12%)",
    borderRadius: "3px !important",
  },
  ".label": {
    textTransform: "none",
  },
});
