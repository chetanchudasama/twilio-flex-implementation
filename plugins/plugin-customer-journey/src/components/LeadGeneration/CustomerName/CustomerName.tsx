import React, { useState } from "react";
import { Grid, Typography, Divider } from "@material-ui/core";
import { CustomInput } from "@common/components";
import { ToggleButton, ToggleButtonGroup } from "@material-ui/lab";

import { CustomerNameStyles } from "./CustomerName.Styles";
import { CustomerNameProps } from "./CustomerName.Props";
import { PreferredNameType } from "../../../common/enum";

export const CustomerName: React.FC<CustomerNameProps> = ({
  updateCustomerPreferredName,
}) => {
  const [selectedSwitch, setSelectedSwitch] = useState<PreferredNameType>(
    PreferredNameType.FullName
  );
  const [preferredName, setPreferredName] = useState<string>("");

  const onThreeStateSwitchChange = (
    event: React.MouseEvent<HTMLElement>,
    value: PreferredNameType
  ) => {
    setPreferredName("");
    setSelectedSwitch(value || PreferredNameType.FullName);
  };

  const onPreferredNameChange = (value: string) => {
    setPreferredName(value);
  };

  const handleUpdatePreferredName = (value: string) => {
    updateCustomerPreferredName(value);
  };

  return (
    <CustomerNameStyles>
      <Grid container spacing={16}>
        <Grid item xs={12} className="heading">
          Customer name
        </Grid>
        <Grid item xs={12} className="customer-name-toggle">
          <Grid
            container
            direction="row"
            justify="space-between"
            alignItems="center"
            spacing={8}
          >
            <Grid item>
              <Typography component="p" variant="caption">
                How does the customer like to be addressed?
              </Typography>
            </Grid>
            <Grid item>
              <ToggleButtonGroup
                value={selectedSwitch}
                exclusive
                onChange={onThreeStateSwitchChange}
                selected
                className="root"
              >
                <ToggleButton
                  className={`button label btn-full-name ${
                    selectedSwitch === PreferredNameType.FullName
                      ? "selected"
                      : ""
                  }`}
                  value={PreferredNameType.FullName}
                  selected={selectedSwitch === PreferredNameType.FullName}
                >
                  Full Name
                </ToggleButton>
                <ToggleButton
                  className={`button label btn-first-name ${
                    selectedSwitch === PreferredNameType.FirstName
                      ? "selected"
                      : ""
                  }`}
                  value={PreferredNameType.FirstName}
                  selected={selectedSwitch === PreferredNameType.FirstName}
                >
                  First Name
                </ToggleButton>
                <ToggleButton
                  className={`button label btn-other ${
                    selectedSwitch === PreferredNameType.Other ? "selected" : ""
                  }`}
                  value={PreferredNameType.Other}
                  selected={selectedSwitch === PreferredNameType.Other}
                >
                  Other
                </ToggleButton>
              </ToggleButtonGroup>
            </Grid>
            {selectedSwitch === PreferredNameType.Other && (
              <Grid item xs={12}>
                <Divider />
              </Grid>
            )}
          </Grid>
        </Grid>
        {selectedSwitch === PreferredNameType.Other && (
          <Grid item xs={12} className="custom-name">
            <Grid
              container
              direction="row"
              justify="space-between"
              alignItems="center"
            >
              <Grid item>
                <Typography component="p" variant="caption">
                  What name do they like to be addressed by?
                </Typography>
              </Grid>
              <Grid item>
                <CustomInput
                  label=""
                  value={preferredName}
                  onChange={(event) =>
                    onPreferredNameChange(event.target.value)
                  }
                  classes={{
                    formControl: "custom-input",
                  }}
                  onBlur={() => handleUpdatePreferredName(preferredName)}
                />
              </Grid>
            </Grid>
          </Grid>
        )}
      </Grid>
    </CustomerNameStyles>
  );
};
