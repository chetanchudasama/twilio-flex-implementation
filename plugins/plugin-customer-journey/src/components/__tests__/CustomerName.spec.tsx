import React from "react";
import { mount } from "enzyme";
import { create } from "react-test-renderer";

import { CustomerNameProps } from "../LeadGeneration/CustomerName/CustomerName.Props";
import { CustomerName } from "../LeadGeneration/CustomerName/CustomerName";

const props: CustomerNameProps = {
  updateCustomerPreferredName: jest.fn(),
  firstName: "john",
  fullName: "Mr John Fred Smith",
};

describe("CustomerName", () => {
  it("renders correctly on load", async () => {
    const wrapper = create(<CustomerName {...props} />).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it("should call updateCustomerPreferredName method, when preferred name textbox blur event is called", () => {
    const wrapper = mount(<CustomerName {...props} />);

    wrapper.find(".btn-other").last().simulate("click");
    wrapper.update();

    const input = wrapper.find("input").last();
    input.simulate("blur");
    wrapper.update();

    expect(props.updateCustomerPreferredName).toHaveBeenCalled();
  });
});
