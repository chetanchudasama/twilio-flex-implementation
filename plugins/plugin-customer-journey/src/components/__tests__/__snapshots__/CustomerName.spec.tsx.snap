// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerName renders correctly on load 1`] = `
<div
  className="css-1lqaj41"
>
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-16-24"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 heading"
    >
      Customer name
    </div>
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 customer-name-toggle"
    >
      <div
        className="MuiGrid-container-1 MuiGrid-spacing-xs-8-23 MuiGrid-align-items-xs-center-9 MuiGrid-justify-xs-space-between-20"
      >
        <div
          className="MuiGrid-item-2"
        >
          <p
            className="MuiTypography-root-98 MuiTypography-caption-108"
          >
            How does the customer like to be addressed?
          </p>
        </div>
        <div
          className="MuiGrid-item-2"
        >
          <div
            className="MuiToggleButtonGroup-root-134 MuiToggleButtonGroup-selected-135 root"
          >
            <button
              className="MuiButtonBase-root-140 MuiToggleButton-root-136 MuiToggleButton-selected-138 button label btn-full-name selected"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex="0"
              type="button"
              value={0}
            >
              <span
                className="MuiToggleButton-label-139"
              >
                Full Name
              </span>
              <span
                className="MuiTouchRipple-root-143"
              />
            </button>
            <button
              className="MuiButtonBase-root-140 MuiToggleButton-root-136 button label btn-first-name "
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex="0"
              type="button"
              value={1}
            >
              <span
                className="MuiToggleButton-label-139"
              >
                First Name
              </span>
              <span
                className="MuiTouchRipple-root-143"
              />
            </button>
            <button
              className="MuiButtonBase-root-140 MuiToggleButton-root-136 button label btn-other "
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex="0"
              type="button"
              value={2}
            >
              <span
                className="MuiToggleButton-label-139"
              >
                Other
              </span>
              <span
                className="MuiTouchRipple-root-143"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
